content/blog/2016/2016-06-29-from-freestyle-to-pipeline.adoc

```diff
@@ -0,0 +1,213 @@
+This statistics generation process primarily performs two major tasks, on rather
+large sets of data:
+
+. Generate aggregate monthly "census data."
+. Process the census data and create trend charts
+
+
+The chained jobs allowed us to resume the independent stages of the pipeline,
+and allowed us to run different stages on different hardware (different
+capabilities) as needed. Below is a diagram of what this looked like:
+
+image::/images/post-images/freestyle-to-pipeline-2016/freestyle-pipeline.png[role=center]
+
+
+The `infra_generate_monthly_json` would run periodically creating the
+aggregated census data, which would then be picked up by `infra_census_push`
+whose sole responsibility was to take census data and publish it to the
+necessary hosts inside the project's infrastructure.
+
+
+The second, semi-independent, "pipeline" would also run periodically. The
+`infra_statistics` job's responsibility was to use the census data, which had
+been previously pushed, to generate the myriad of pretty blue charts before notifying the
+`infra_checkout_stats` job which would make sure `stats.jenkins-ci.org` was
+properly updated.
+
+
+Suffice it to say, this "pipeline" had grown organically over a period time when
+link:/doc/pipeline[more advanced tools] weren't quite available.
+
+
+---
+
+
+When we migrated to newer infrastructure for
+link:https://ci.jenkins.io[ci.jenkins.io] earlier this year, instead of
+migrating jobs verbatim, I took the opportunity to prune stale jobs and refactor
+a number of others into proper link:/solutions/pipeline[Pipelines], statistics
+generation being an obvious target!
+
+Our requirements for statistics generation, in their most basic form, are:
+
+* Enable a sequence of dependent tasks to be executed as a logical group (a
+  pipeline)
+* Enable executing those dependent tasks on various pieces of infrastructure
+  which support different requirements
+* Actually generate those pretty blue charts
+
+
+[NOTE]
+--
+If you wish to skip ahead, you can jump straight to the
+link:https://github.com/jenkins-infra/infra-statistics/blob/a6dcaa29fca9a4f61143954fb9e1300c2f995a89/Jenkinsfile[Jenkinsfile]
+which implements our new Pipeline.
+--
+
+
+The first iteration of the `Jenkinsfile` simply defined the conceptual stages we
+would need:
+
+[source, groovy]
+----
+
+node {
+    stage 'Sync raw data and census files'
+
+    stage 'Process raw logs'
+
+    stage 'Generate census data'
+
+    stage 'Generate stats'
+
+    stage 'Publish census'
+
+    stage 'Publish stats'
+}
+----
+
+How exciting! But not terrifically useful. With the implementation of the first
+couple stages I noticed that the Pipeline might be syncing _dozens_ of gigabytes
+of data every time it ran on a new agent in the cluster. While this problem
+will soon be solved by the
+link:https://github.com/jenkinsci/external-workspace-manager-plugin[External
+Workspace Manager plugin], which is currently being developed, I needed to pin
```

Comment:
"developed, I needed to pin" ->
"developed. Until then, I chose to mitigate the issue by pinning"